File: bindings/csharp/Scope.cs
using Microsoft.Win32.SafeHandles;

namespace Pap;

/// <summary>Deny-by-default set of permitted Schema.org actions.</summary>
public sealed class Scope : IDisposable
{
    private sealed class Handle : SafeHandleZeroOrMinusOneIsInvalid
    {
        internal Handle(IntPtr h) : base(ownsHandle: true) { SetHandle(h); }
        protected override bool ReleaseHandle() { PapNative.pap_scope_free(handle); return true; }
    }

    private readonly Handle _handle;
    internal IntPtr Raw => _handle.DangerousGetHandle();

    private Scope(IntPtr h) => _handle = new Handle(h);

    /// <summary>
    /// Build a scope from an array of (action, optional-object) tuples.
    /// Passing an empty array produces a deny-all scope.
    /// </summary>
    public static Scope From(params (string Action, string? Object)[] actions)
    {
        // Build and accumulate C-side action handles.
        var handles = new IntPtr[actions.Length];
        for (int i = 0; i < actions.Length; i++)
        {
            var (action, obj) = actions[i];
            handles[i] = obj is null
                ? PapNative.pap_scope_action_new(action)
                : PapNative.pap_scope_action_with_object(action, obj);

            if (handles[i] == IntPtr.Zero)
            {
                // Free already-created handles before throwing.
                for (int j = 0; j < i; j++) PapNative.pap_scope_action_free(handles[j]);
                PapNative.ThrowLastError("scope_action_new");
            }
        }

        // Build the Scope (clones the actions internally).
        var h = PapNative.pap_scope_new(handles, (nuint)handles.Length);
        foreach (var a in handles) PapNative.pap_scope_action_free(a);
        if (h == IntPtr.Zero) PapNative.ThrowLastError("scope_new");
        return new Scope(h);
    }

    /// <summary>Create a deny-all scope.</summary>
    public static Scope DenyAll()
    {
        var h = PapNative.pap_scope_deny_all();
        if (h == IntPtr.Zero) PapNative.ThrowLastError("scope_deny_all");
        return new Scope(h);
    }

    /// <summary>Returns <c>true</c> if the scope permits <paramref name="action"/>.</summary>
    public bool Permits(string action) =>
        PapNative.pap_scope_permits(Raw, action) == 1;

    /// <summary>Returns <c>true</c> if <paramref name="child"/> ⊆ this scope.</summary>
    public bool Contains(Scope child) =>
        PapNative.pap_scope_contains(Raw, child.Raw) == 1;

    public void Dispose() => _handle.Dispose();
}

File: bindings/java/src/main/java/io/pap/Scope.java
package io.pap;

import com.sun.jna.Pointer;

/**
 * Deny-by-default set of permitted Schema.org actions.
 * An agent can only perform actions explicitly listed in its mandate scope.
 */
public final class Scope implements AutoCloseable {

    private final Pointer handle;

    private Scope(Pointer h) {
        this.handle = h;
    }

    /**
     * Build a scope from a flat list of action names (no object constraints).
     * This is the common case — each string is one permitted Schema.org action.
     *
     * <pre>{@code
     * Scope s = Scope.from(new String[]{
     *     "schema:SearchAction",
     *     "schema:ReserveAction"
     * });
     * }</pre>
     *
     * An empty array creates a deny-all scope (nothing permitted).
     */
    public static Scope from(String[] actions) {
        String[][] pairs = new String[actions.length][];
        for (int i = 0; i < actions.length; i++) {
            pairs[i] = new String[]{actions[i]};
        }
        return from(pairs);
    }

    /**
     * Build a scope from a list of (action, optional-object) pairs.
     * Use {@code null} for the object when there is no constraint.
     *
     * <pre>{@code
     * Scope s = Scope.from(
     *     new String[]{"schema:SearchAction", null},
     *     new String[]{"schema:ReserveAction", "schema:Flight"}
     * );
     * }</pre>
     */
    public static Scope from(String[]... actionPairs) {
        Pointer[] actionHandles = new Pointer[actionPairs.length];
        for (int i = 0; i < actionPairs.length; i++) {
            String action = actionPairs[i][0];
            String object = actionPairs[i].length > 1 ? actionPairs[i][1] : null;
            Pointer a = object == null
                    ? PapLib.INSTANCE.pap_scope_action_new(action)
                    : PapLib.INSTANCE.pap_scope_action_with_object(action, object);
            if (a == null) {
                for (int j = 0; j < i; j++) PapLib.INSTANCE.pap_scope_action_free(actionHandles[j]);
                throw PapException.fromLastError("scope_action_new");
            }
            actionHandles[i] = a;
        }
        Pointer h = PapLib.INSTANCE.pap_scope_new(actionHandles, actionHandles.length);
        for (Pointer a : actionHandles) PapLib.INSTANCE.pap_scope_action_free(a);
        if (h == null) throw PapException.fromLastError("scope_new");
        return new Scope(h);
    }

    /** Create a deny-all scope. */
    public static Scope denyAll() {
        Pointer h = PapLib.INSTANCE.pap_scope_deny_all();
        if (h == null) throw PapException.fromLastError("scope_deny_all");
        return new Scope(h);
    }

    /** Returns {@code true} if the scope permits {@code action}. */
    public boolean permits(String action) {
        return PapLib.INSTANCE.pap_scope_permits(handle, action) == 1;
    }

    /** Returns {@code true} if {@code child} ⊆ this scope. */
    public boolean contains(Scope child) {
        return PapLib.INSTANCE.pap_scope_contains(handle, child.handle) == 1;
    }

    Pointer raw() { return handle; }

    @Override
    public void close() {
        PapLib.INSTANCE.pap_scope_free(handle);
    }
}
